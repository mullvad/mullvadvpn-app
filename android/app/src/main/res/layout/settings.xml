<FrameLayout xmlns:android="http://schemas.android.com/apk/res/android"
             xmlns:mullvad="http://schemas.android.com/apk/res-auto"
             android:layout_width="match_parent"
             android:layout_height="match_parent"
             android:background="@color/darkBlue"
             android:gravity="left">
    <TextView android:id="@+id/title"
              android:layout_width="wrap_content"
              android:layout_height="wrap_content"
              android:text="@string/settings"
              style="@style/SettingsCollapsedHeader" />
    <LinearLayout android:layout_width="match_parent"
                  android:layout_height="match_parent"
                  android:orientation="vertical">
        <FrameLayout android:layout_width="match_parent"
                     android:layout_height="wrap_content">
            <ImageButton android:id="@+id/close"
                         android:layout_width="wrap_content"
                         android:layout_height="wrap_content"
                         android:padding="12dp"
                         android:background="?android:attr/selectableItemBackground"
                         android:src="@drawable/icon_close" />
            <TextView android:id="@+id/collapsed_title"
                      android:layout_width="wrap_content"
                      android:layout_height="wrap_content"
                      android:layout_marginHorizontal="4dp"
                      android:layout_gravity="center"
                      android:text="@string/settings"
                      style="@style/SettingsCollapsedHeader" />
        </FrameLayout>
        <net.mullvad.mullvadvpn.ui.widget.ListenableScrollView android:id="@+id/scroll_area"
                                                               android:layout_width="match_parent"
                                                               android:layout_height="match_parent">

            <LinearLayout android:layout_width="match_parent"
                          android:layout_height="wrap_content"
                          android:orientation="vertical">
                <TextView android:id="@+id/expanded_title"
                          android:layout_width="wrap_content"
                          android:layout_height="wrap_content"
                          android:layout_marginTop="4dp"
                          android:layout_marginLeft="@dimen/side_margin"
                          android:lines="1"
                          android:text="@string/settings"
                          style="@style/SettingsExpandedHeader" />
                <net.mullvad.mullvadvpn.ui.widget.AccountCell android:id="@+id/account"
                                                              android:layout_width="match_parent"
                                                              android:layout_height="wrap_content"
                                                              android:layout_marginTop="@dimen/vertical_space"
                                                              mullvad:text="@string/settings_account" />
                <net.mullvad.mullvadvpn.ui.widget.NavigateCell android:id="@+id/preferences"
                                                               android:layout_width="match_parent"
                                                               android:layout_height="wrap_content"
                                                               android:layout_marginTop="1dp"
                                                               mullvad:text="@string/settings_preferences" />
                <net.mullvad.mullvadvpn.ui.widget.NavigateCell android:id="@+id/advanced"
                                                               android:layout_width="match_parent"
                                                               android:layout_height="wrap_content"
                                                               android:layout_marginTop="1dp"
                                                               mullvad:text="@string/settings_advanced" />
                <net.mullvad.mullvadvpn.ui.widget.AppVersionCell android:id="@+id/app_version"
                                                                 android:layout_width="match_parent"
                                                                 android:layout_height="wrap_content"
                                                                 android:layout_marginTop="@dimen/vertical_space"
                                                                 mullvad:text="@string/app_version"
                                                                 mullvad:footer="@string/update_available_footer" />
                <net.mullvad.mullvadvpn.ui.widget.NavigateCell android:id="@+id/report_a_problem"
                                                               android:layout_width="match_parent"
                                                               android:layout_height="wrap_content"
                                                               android:layout_marginTop="@dimen/vertical_space"
                                                               mullvad:text="@string/report_a_problem" />
                <net.mullvad.mullvadvpn.ui.widget.UrlCell android:id="@+id/faqs_and_guides"
                                                          android:layout_width="match_parent"
                                                          android:layout_height="wrap_content"
                                                          android:layout_marginTop="1dp"
                                                          mullvad:text="@string/faqs_and_guides"
                                                          mullvad:url="@string/faqs_and_guides_url" />
                <net.mullvad.mullvadvpn.ui.widget.UrlCell android:layout_width="match_parent"
                                                          android:layout_height="wrap_content"
                                                          android:layout_marginTop="1dp"
                                                          mullvad:text="@string/privacy_policy"
                                                          mullvad:url="@string/privacy_policy_url" />
            </LinearLayout>
        </net.mullvad.mullvadvpn.ui.widget.ListenableScrollView>
    </LinearLayout>
</FrameLayout>
