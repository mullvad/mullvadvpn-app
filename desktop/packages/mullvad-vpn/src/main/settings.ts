import fs from 'fs/promises';

import { ISettings } from '../shared/daemon-rpc-types';
import { ICurrentAppVersionInfo } from '../shared/ipc-types';
import log from '../shared/logging';
import { getOpenAtLogin, setOpenAtLogin } from './autostart';
import { DaemonRpc } from './daemon-rpc';
import { getDefaultSettings } from './default-settings';
import GuiSettings from './gui-settings';
import { IpcMainEventChannel } from './ipc-event-channel';

export interface SettingsDelegate {
  handleMonochromaticIconChange(value: boolean): void;
  handleUnpinnedWindowChange(): void;
}

export default class Settings implements Readonly<ISettings> {
  private guiSettings = new GuiSettings();

  private settingsValue = getDefaultSettings();

  public constructor(
    private delegate: SettingsDelegate,
    private daemonRpc: DaemonRpc,
    private currentVersion: ICurrentAppVersionInfo,
  ) {}

  public registerIpcListeners() {
    this.registerGuiSettingsListener();

    IpcMainEventChannel.settings.handleSetAllowLan((allowLan) =>
      this.daemonRpc.setAllowLan(allowLan),
    );
    IpcMainEventChannel.settings.handleSetShowBetaReleases((showBetaReleases) =>
      this.daemonRpc.setShowBetaReleases(showBetaReleases),
    );
    IpcMainEventChannel.settings.handleSetEnableIpv6((enableIpv6) =>
      this.daemonRpc.setEnableIpv6(enableIpv6),
    );
    IpcMainEventChannel.settings.handleSetBlockWhenDisconnected((blockWhenDisconnected) =>
      this.daemonRpc.setBlockWhenDisconnected(blockWhenDisconnected),
    );
    IpcMainEventChannel.settings.handleSetBridgeState(async (bridgeState) => {
      await this.daemonRpc.setBridgeState(bridgeState);

      // Reset bridge constraints to `any` when the state is set to auto or off if not custom
      if (
        (bridgeState === 'auto' || bridgeState === 'off') &&
        this.bridgeSettings.type === 'normal'
      ) {
        await this.daemonRpc.setBridgeSettings({
          ...this.bridgeSettings,
          normal: { ...this.bridgeSettings.normal, location: 'any' },
        });
      }
    });
    IpcMainEventChannel.settings.handleSetOpenVpnMssfix((mssfix?: number) =>
      this.daemonRpc.setOpenVpnMssfix(mssfix),
    );
    IpcMainEventChannel.settings.handleSetWireguardMtu((mtu?: number) =>
      this.daemonRpc.setWireguardMtu(mtu),
    );
    IpcMainEventChannel.settings.handleSetWireguardQuantumResistant((quantumResistant?: boolean) =>
      this.daemonRpc.setWireguardQuantumResistant(quantumResistant),
    );
    IpcMainEventChannel.settings.handleSetRelaySettings((relaySettings) =>
      this.daemonRpc.setRelaySettings(relaySettings),
    );
    IpcMainEventChannel.settings.handleUpdateBridgeSettings((bridgeSettings) => {
      return this.daemonRpc.setBridgeSettings(bridgeSettings);
    });
    IpcMainEventChannel.settings.handleSetDnsOptions((dns) => {
      return this.daemonRpc.setDnsOptions(dns);
    });
    IpcMainEventChannel.autoStart.handleSet((autoStart: boolean) => {
      return this.setAutoStart(autoStart);
    });
    IpcMainEventChannel.settings.handleSetObfuscationSettings((obfuscationSettings) => {
      return this.daemonRpc.setObfuscationSettings(obfuscationSettings);
    });
    IpcMainEventChannel.settings.handleAddApiAccessMethod((method) => {
      return this.daemonRpc.addApiAccessMethod(method);
    });
    IpcMainEventChannel.settings.handleUpdateApiAccessMethod((method) => {
      return this.daemonRpc.updateApiAccessMethod(method);
    });
    IpcMainEventChannel.settings.handleRemoveApiAccessMethod((id) => {
      return this.daemonRpc.removeApiAccessMethod(id);
    });
    IpcMainEventChannel.settings.handleSetApiAccessMethod((id) => {
      return this.daemonRpc.setApiAccessMethod(id);
    });
    IpcMainEventChannel.settings.handleTestApiAccessMethodById((id) => {
      return this.daemonRpc.testApiAccessMethodById(id);
    });
    IpcMainEventChannel.settings.handleTestCustomApiAccessMethod((method) => {
      return this.daemonRpc.testCustomApiAccessMethod(method);
    });

    IpcMainEventChannel.settings.handleClearAllRelayOverrides(() => {
      return this.daemonRpc.clearAllRelayOverrides();
    });
    IpcMainEventChannel.settings.handleImportText((text) => {
      return this.daemonRpc.applyJsonSettings(text);
    });
    IpcMainEventChannel.settings.handleImportFile(async (path) => {
      const settings = await fs.readFile(path);
      return this.daemonRpc.applyJsonSettings(settings.toString());
    });
    IpcMainEventChannel.settings.handleSetEnableDaita((value) => {
      return this.daemonRpc.setEnableDaita(value);
    });
    IpcMainEventChannel.settings.handleSetDaitaDirectOnly((value) => {
      return this.daemonRpc.setDaitaDirectOnly(value);
    });

    IpcMainEventChannel.guiSettings.handleSetEnableSystemNotifications((flag: boolean) => {
      this.guiSettings.enableSystemNotifications = flag;
    });

    IpcMainEventChannel.guiSettings.handleSetAutoConnect((autoConnect: boolean) => {
      this.guiSettings.autoConnect = autoConnect;
    });

    IpcMainEventChannel.guiSettings.handleSetStartMinimized((startMinimized: boolean) => {
      this.guiSettings.startMinimized = startMinimized;
    });

    IpcMainEventChannel.guiSettings.handleSetMonochromaticIcon((monochromaticIcon: boolean) => {
      this.guiSettings.monochromaticIcon = monochromaticIcon;
    });

    IpcMainEventChannel.guiSettings.handleSetUnpinnedWindow((unpinnedWindow: boolean) => {
      this.guiSettings.unpinnedWindow = unpinnedWindow;
      this.delegate.handleUnpinnedWindowChange();
    });

    IpcMainEventChannel.guiSettings.handleSetAnimateMap((animateMap: boolean) => {
      this.guiSettings.animateMap = animateMap;
    });

    IpcMainEventChannel.currentVersion.handleDisplayedChangelog(() => {
      this.guiSettings.changelogDisplayedForVersion = this.currentVersion.gui;
    });

    IpcMainEventChannel.upgradeVersion.handleDismissedUpgrade((version: string) => {
      this.guiSettings.updateDismissedForVersion = version;
    });
  }

  public get all() {
    return this.settingsValue;
  }

  public get allowLan() {
    return this.settingsValue.allowLan;
  }
  public get autoConnect() {
    return this.settingsValue.autoConnect;
  }
  public get blockWhenDisconnected() {
    return this.settingsValue.blockWhenDisconnected;
  }
  public get showBetaReleases() {
    return this.settingsValue.showBetaReleases;
  }
  public get relaySettings() {
    return this.settingsValue.relaySettings;
  }
  public get tunnelOptions() {
    return this.settingsValue.tunnelOptions;
  }
  public get bridgeSettings() {
    return this.settingsValue.bridgeSettings;
  }
  public get bridgeState() {
    return this.settingsValue.bridgeState;
  }
  public get splitTunnel() {
    return this.settingsValue.splitTunnel;
  }
  public get obfuscationSettings() {
    return this.settingsValue.obfuscationSettings;
  }
  public get customLists() {
    return this.settingsValue.customLists;
  }
  public get apiAccessMethods() {
    return this.settingsValue.apiAccessMethods;
  }
  public get relayOverrides() {
    return this.settingsValue.relayOverrides;
  }

  public get gui() {
    return this.guiSettings;
  }

  public handleNewSettings(newSettings: ISettings) {
    this.settingsValue = newSettings;
  }

  private registerGuiSettingsListener() {
    this.guiSettings.onChange = (newState, oldState) => {
      if (oldState.monochromaticIcon !== newState.monochromaticIcon) {
        this.delegate.handleMonochromaticIconChange(newState.monochromaticIcon);
      }

      if (newState.autoConnect !== oldState.autoConnect) {
        this.updateDaemonsAutoConnect();
      }

      IpcMainEventChannel.guiSettings.notify?.(newState);
    };
  }

  private async setAutoStart(autoStart: boolean): Promise<void> {
    try {
      await setOpenAtLogin(autoStart);

      IpcMainEventChannel.autoStart.notify?.(autoStart);

      this.updateDaemonsAutoConnect();
    } catch (e) {
      const error = e as Error;
      log.error(
        `Failed to update the autostart to ${autoStart.toString()}. ${error.message.toString()}`,
      );
    }
    return Promise.resolve();
  }

  private updateDaemonsAutoConnect() {
    const daemonAutoConnect = this.guiSettings.autoConnect && getOpenAtLogin();
    if (daemonAutoConnect !== this.settingsValue.autoConnect) {
      void this.daemonRpc.setAutoConnect(daemonAutoConnect);
    }
  }
}
