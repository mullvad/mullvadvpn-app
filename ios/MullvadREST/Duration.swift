//
//  Duration.swift
//  MullvadREST
//
//  Created by pronebird on 04/11/2022.
//  Copyright Â© 2022 Mullvad VPN AB. All rights reserved.
//

import Foundation

extension REST {
    public struct Duration: Comparable {
        public let milliseconds: UInt64

        public var seconds: UInt64 {
            return milliseconds / 1000
        }

        public var timeInterval: TimeInterval {
            return TimeInterval(milliseconds) / 1000
        }

        private init(milliseconds: UInt64) {
            self.milliseconds = milliseconds
        }

        public func format() -> String {
            guard milliseconds >= 1000 else {
                return "\(milliseconds)ms"
            }

            let trailingZeroesSuffix = ".00"
            var string = String(format: "%.2f", timeInterval)

            if string.hasSuffix(trailingZeroesSuffix) {
                string.removeLast(trailingZeroesSuffix.count)
            }

            return "\(string)s"
        }

        public static func seconds(_ seconds: UInt64) -> Duration {
            return Duration(milliseconds: seconds.saturatingMultiplication(1000))
        }

        public static func milliseconds(_ milliseconds: UInt64) -> Duration {
            return Duration(milliseconds: milliseconds)
        }

        public static func < (lhs: Duration, rhs: Duration) -> Bool {
            return lhs.milliseconds < rhs.milliseconds
        }

        public static func == (lhs: Duration, rhs: Duration) -> Bool {
            return lhs.milliseconds == rhs.milliseconds
        }

        public static func * (lhs: Duration, factor: UInt64) -> Duration {
            return Duration(milliseconds: lhs.milliseconds.saturatingMultiplication(factor))
        }
    }
}
