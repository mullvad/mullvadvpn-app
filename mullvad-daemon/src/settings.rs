#[cfg(not(target_os = "android"))]
use futures::TryFutureExt;
use mullvad_types::{
    relay_constraints::{RelayConstraints, RelaySettings, WireguardConstraints},
    settings::{DnsState, Settings},
};
use rand::Rng;
use std::{
    fmt::{self, Display},
    ops::Deref,
    path::{Path, PathBuf},
};
use talpid_core::firewall::is_local_address;
use talpid_types::ErrorExt;
use tokio::{
    fs,
    io::{self, AsyncWriteExt},
};

const SETTINGS_FILE: &str = "settings.json";

#[derive(err_derive::Error, Debug)]
#[error(no_from)]
pub enum Error {
    #[error(display = "Unable to read settings file {}", _0)]
    ReadError(String, #[error(source)] io::Error),

    #[error(display = "Unable to parse settings file")]
    ParseError(#[error(source)] serde_json::Error),

    #[error(display = "Unable to remove settings file {}", _0)]
    #[cfg(not(target_os = "android"))]
    DeleteError(String, #[error(source)] io::Error),

    #[error(display = "Unable to serialize settings to JSON")]
    SerializeError(#[error(source)] serde_json::Error),

    #[error(display = "Unable to write settings to {}", _0)]
    WriteError(String, #[error(source)] io::Error),
}

#[derive(Debug)]
pub struct SettingsPersister {
    settings: Settings,
    path: PathBuf,
}

pub type MadeChanges = bool;

impl SettingsPersister {
    /// Loads user settings from file. If it fails, it returns the defaults.
    pub async fn load(settings_dir: &Path) -> Self {
        let path = settings_dir.join(SETTINGS_FILE);
        let (mut settings, mut should_save) = match Self::load_from_file(&path).await {
            Ok(value) => value,
            Err(error) => {
                log::warn!(
                    "{}",
                    error.display_chain_with_msg("Failed to load settings. Using defaults.")
                );
                let mut settings = Self::default_settings();

                // Protect the user by blocking the internet by default. Previous settings may
                // not have caused the daemon to enter the non-blocking disconnected state.
                settings.block_when_disconnected = true;

                (settings, true)
            }
        };

        // If the settings file did not contain a wg_migration_rand_num then it will be initialized
        // to -1.0 by serde. This block ensures that this value is correctly intitialzed to a
        // percentage.
        if settings.wg_migration_rand_num < 0.0 || settings.wg_migration_rand_num > 1.0 {
            let mut rng = rand::thread_rng();
            settings.wg_migration_rand_num = rng.gen_range(0.0..=1.0);
            should_save |= true
        }

        // Force IPv6 to be enabled on Android
        if cfg!(target_os = "android") {
            should_save |= !settings.tunnel_options.generic.enable_ipv6;
            settings.tunnel_options.generic.enable_ipv6 = true;
        }
        if crate::version::is_beta_version() {
            should_save |= !settings.show_beta_releases;
            settings.show_beta_releases = true;
        }

        let mut persister = SettingsPersister { settings, path };

        if should_save {
            if let Err(error) = persister.save().await {
                log::error!(
                    "{}",
                    error.display_chain_with_msg("Failed to save updated settings")
                );
            }
        }

        persister
    }

    async fn load_from_file(path: &Path) -> Result<(Settings, bool), Error> {
        log::info!("Loading settings from {}", path.display());

        let settings_bytes = match fs::read(path).await {
            Ok(bytes) => bytes,
            Err(error) => {
                if error.kind() == io::ErrorKind::NotFound {
                    log::info!("No settings were found. Using defaults.");
                    return Ok((Self::default_settings(), true));
                } else {
                    return Err(Error::ReadError(path.display().to_string(), error));
                }
            }
        };
        Ok((Self::load_from_bytes(&settings_bytes)?, false))
    }

    fn load_from_bytes(bytes: &[u8]) -> Result<Settings, Error> {
        serde_json::from_slice(bytes).map_err(Error::ParseError)
    }

    async fn save(&mut self) -> Result<(), Error> {
        Self::save_inner(&self.path, &self.settings).await
    }

    /// Serializes the settings and saves them to the given file.
    async fn save_inner(path: &Path, settings: &Settings) -> Result<(), Error> {
        log::debug!("Writing settings to {}", path.display());

        let buffer = serde_json::to_string_pretty(settings).map_err(Error::SerializeError)?;
        let mut file = mullvad_fs::AtomicFile::new(path)
            .await
            .map_err(|e| Error::WriteError(path.display().to_string(), e))?;
        file.write_all(&buffer.into_bytes())
            .await
            .map_err(|e| Error::WriteError(path.display().to_string(), e))?;
        file.finalize()
            .await
            .map_err(|e| Error::WriteError(path.display().to_string(), e))?;

        Ok(())
    }

    /// Resets default settings
    #[cfg(not(target_os = "android"))]
    pub async fn reset(&mut self) -> Result<(), Error> {
        self.settings = Self::default_settings();
        let path = self.path.clone();
        self.save()
            .or_else(|e| async move {
                log::error!(
                    "{}",
                    e.display_chain_with_msg("Unable to save default settings")
                );
                log::info!("Will attempt to remove settings file");
                fs::remove_file(&path)
                    .map_err(|e| Error::DeleteError(path.display().to_string(), e))
                    .await
            })
            .await
    }

    pub fn to_settings(&self) -> Settings {
        self.settings.clone()
    }

    /// Modifies `Settings::default()` somewhat, e.g. depending on whether a beta version
    /// is being run or not.
    fn default_settings() -> Settings {
        let mut settings = Settings::default();

        if crate::version::is_beta_version() {
            settings.show_beta_releases = true;
        }
        settings
    }

    /// Edit the settings in a closure, and write the changes, if any, to disk.
    ///
    /// On success, the function returns a boolean indicating whether any settings were changed.
    /// If the settings could not be written to disk, all changes are rolled back, and an error is
    /// returned.
    pub async fn update(
        &mut self,
        update_fn: impl FnOnce(&mut Settings),
    ) -> Result<MadeChanges, Error> {
        let mut new_settings = self.settings.clone();

        update_fn(&mut new_settings);

        if self.settings == new_settings {
            return Ok(false);
        }

        Self::save_inner(&self.path, &new_settings).await?;
        self.settings = new_settings;
        Ok(true)
    }

    /// Return a compact summary of important settings
    pub fn summary(&self) -> SettingsSummary<'_> {
        SettingsSummary {
            settings: &self.settings,
        }
    }
}

impl Deref for SettingsPersister {
    type Target = Settings;

    fn deref(&self) -> &Self::Target {
        &self.settings
    }
}

/// A compact summary of important settings
pub struct SettingsSummary<'a> {
    settings: &'a Settings,
}

impl<'a> Display for SettingsSummary<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let bool_to_label = |state| {
            if state {
                "on"
            } else {
                "off"
            }
        };

        let relay_settings = self.settings.get_relay_settings();

        write!(f, "ovpn mssfix: ")?;
        Self::fmt_option(f, self.settings.tunnel_options.openvpn.mssfix)?;
        write!(f, ", wg mtu: ")?;
        Self::fmt_option(f, self.settings.tunnel_options.wireguard.mtu)?;

        if let RelaySettings::Normal(RelayConstraints {
            wireguard_constraints: WireguardConstraints { ip_version, .. },
            ..
        }) = relay_settings
        {
            write!(f, ", wg ip: {ip_version}")?;
        }

        let multihop = matches!(
            relay_settings,
            RelaySettings::Normal(RelayConstraints {
                wireguard_constraints: WireguardConstraints {
                    use_multihop: true,
                    ..
                },
                ..
            })
        );

        write!(
            f,
            ", multihop: {}, ipv6 (tun): {}, lan: {}, pq: {}, obfs: {}",
            bool_to_label(multihop),
            bool_to_label(self.settings.tunnel_options.generic.enable_ipv6),
            bool_to_label(self.settings.allow_lan),
            self.settings.tunnel_options.wireguard.quantum_resistant,
            self.settings.obfuscation_settings.selected_obfuscation,
        )?;

        // Print DNS options

        match self.settings.tunnel_options.dns_options.state {
            DnsState::Default => {
                let mut content = vec!["default"];

                if self
                    .settings
                    .tunnel_options
                    .dns_options
                    .default_options
                    .block_ads
                {
                    content.push("ads");
                }
                if self
                    .settings
                    .tunnel_options
                    .dns_options
                    .default_options
                    .block_trackers
                {
                    content.push("trackers");
                }
                if self
                    .settings
                    .tunnel_options
                    .dns_options
                    .default_options
                    .block_malware
                {
                    content.push("malware");
                }
                if self
                    .settings
                    .tunnel_options
                    .dns_options
                    .default_options
                    .block_adult_content
                {
                    content.push("adult");
                }
                if self
                    .settings
                    .tunnel_options
                    .dns_options
                    .default_options
                    .block_gambling
                {
                    content.push("gambling");
                }

                write!(f, ", dns: {}", content.join(" "))?;
            }
            DnsState::Custom => {
                // NOTE: Technically inaccurate, as the gateway IP is a local IP but isn't treated as one.
                let contains_local = self
                    .settings
                    .tunnel_options
                    .dns_options
                    .custom_options
                    .addresses
                    .iter()
                    .any(is_local_address);
                let contains_public = self
                    .settings
                    .tunnel_options
                    .dns_options
                    .custom_options
                    .addresses
                    .iter()
                    .any(|addr| !is_local_address(addr));

                match (contains_public, contains_local) {
                    (true, true) => f.write_str(", dns: custom, public, local")?,
                    (true, false) => f.write_str(", dns: custom, public")?,
                    (false, false) => f.write_str(", dns: custom, no addrs")?,
                    (false, true) => f.write_str(", dns: custom, local")?,
                }
            }
        }
        Ok(())
    }
}

impl<'a> SettingsSummary<'a> {
    fn fmt_option<T: Display>(f: &mut fmt::Formatter<'_>, val: Option<T>) -> fmt::Result {
        if let Some(inner) = &val {
            inner.fmt(f)
        } else {
            f.write_str("unset")
        }
    }
}

#[cfg(test)]
mod test {
    use super::SettingsPersister;
    use mullvad_types::settings::SettingsVersion;
    use serde_json;

    #[test]
    #[should_panic]
    fn test_deserialization_failure_version_too_small() {
        let _version: SettingsVersion = serde_json::from_str("1").expect("Version too small");
    }

    #[test]
    #[should_panic]
    fn test_deserialization_failure_version_too_big() {
        let _version: SettingsVersion = serde_json::from_str("1000").expect("Version too big");
    }

    #[test]
    fn test_deserialization_success() {
        let _version: SettingsVersion =
            serde_json::from_str("2").expect("Failed to deserialize valid version");
    }

    #[test]
    fn test_serialization_success() {
        let version = SettingsVersion::V2;
        let s = serde_json::to_string(&version).expect("Failed to serialize");
        assert_eq!(s, "2");
    }

    #[test]
    fn test_deserialization() {
        let settings = br#"{
              "account_token": "0000000000000000",
              "relay_settings": {
                "normal": {
                  "location": {
                    "only": {
                      "country": "gb"
                    }
                  },
                  "tunnel_protocol": {
                    "only": "wireguard"
                  },
                  "wireguard_constraints": {
                    "port": "any"
                  },
                  "openvpn_constraints": {
                    "port": "any",
                    "protocol": "any"
                  }
                }
              },
              "bridge_settings": {
                "normal": {
                  "location": "any"
                }
              },
              "bridge_state": "auto",
              "allow_lan": true,
              "block_when_disconnected": false,
              "auto_connect": true,
              "tunnel_options": {
                "openvpn": {
                  "mssfix": null
                },
                "wireguard": {
                  "mtu": null,
                  "rotation_interval": null
                },
                "generic": {
                  "enable_ipv6": true
                }
              },
              "settings_version": 5,
              "show_beta_releases": false
        }"#;

        let _ = SettingsPersister::load_from_bytes(settings).unwrap();
    }
}
