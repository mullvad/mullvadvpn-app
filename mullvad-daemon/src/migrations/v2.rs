#![allow(clippy::identity_op)]
use super::{Error, Result};
use mullvad_types::settings::SettingsVersion;
use std::time::Duration;

// ======================================================
// Section for vendoring types and values that
// this settings version depend on. See `mod.rs`.

pub const MIN_ROTATION_INTERVAL: Duration = Duration::from_secs(1 * 24 * 60 * 60);
pub const MAX_ROTATION_INTERVAL: Duration = Duration::from_secs(7 * 24 * 60 * 60);

// ======================================================

pub fn migrate(settings: &mut serde_json::Value) -> Result<()> {
    if !version_matches(settings) {
        return Ok(());
    }

    log::info!("Migrating settings format to V3");

    // `show_beta_releases` used to be nullable
    if settings
        .get_mut("show_beta_releases")
        .map(|val| val.is_null())
        .unwrap_or(false)
    {
        settings
            .as_object_mut()
            .ok_or(Error::InvalidSettingsContent)?
            .remove("show_beta_releases");
    }

    let automatic_rotation = || -> Option<u64> {
        settings
            .get("tunnel_options")?
            .get("wireguard")?
            .get("automatic_rotation")
            .map(|ivl| ivl.as_u64())?
    }();

    if let Some(interval) = automatic_rotation {
        let new_ivl = match Duration::from_secs(60 * 60 * interval) {
            ivl if ivl < MIN_ROTATION_INTERVAL => {
                log::warn!("Increasing key rotation interval since it is below minimum");
                MIN_ROTATION_INTERVAL
            }
            ivl if ivl > MAX_ROTATION_INTERVAL => {
                log::warn!("Decreasing key rotation interval since it is above maximum");
                MAX_ROTATION_INTERVAL
            }
            ivl => ivl,
        };

        settings["tunnel_options"]["wireguard"]["rotation_interval"] = serde_json::json!(new_ivl);
        settings["tunnel_options"]["wireguard"]
            .as_object_mut()
            .ok_or(Error::InvalidSettingsContent)?
            .remove("automatic_rotation");
    }

    settings["settings_version"] = serde_json::json!(SettingsVersion::V3);

    Ok(())
}

fn version_matches(settings: &serde_json::Value) -> bool {
    settings
        .get("settings_version")
        .map(|version| version == SettingsVersion::V2 as u64)
        .unwrap_or(false)
}

#[cfg(test)]
mod test {
    use super::{migrate, version_matches};

    const V2_SETTINGS: &str = r#"
{
  "account_token": "1234",
  "relay_settings": {
    "normal": {
      "location": {
        "only": {
          "country": "se"
        }
      },
      "tunnel_protocol": "any",
      "wireguard_constraints": {
        "port": "any"
      },
      "openvpn_constraints": {
        "port": {
          "only": 53
        },
        "protocol": {
          "only": "udp"
        }
      }
    }
  },
  "bridge_settings": {
    "normal": {
      "location": "any"
    }
  },
  "bridge_state": "auto",
  "allow_lan": true,
  "block_when_disconnected": false,
  "auto_connect": false,
  "tunnel_options": {
    "openvpn": {
      "mssfix": null
    },
    "wireguard": {
      "mtu": null,
      "automatic_rotation": 10
    },
    "generic": {
      "enable_ipv6": false
    }
  },
  "show_beta_releases": null,
  "settings_version": 2
}
"#;

    pub const V3_SETTINGS: &str = r#"
{
  "account_token": "1234",
  "relay_settings": {
    "normal": {
      "location": {
        "only": {
          "country": "se"
        }
      },
      "tunnel_protocol": "any",
      "wireguard_constraints": {
        "port": "any"
      },
      "openvpn_constraints": {
        "port": {
          "only": 53
        },
        "protocol": {
          "only": "udp"
        }
      }
    }
  },
  "bridge_settings": {
    "normal": {
      "location": "any"
    }
  },
  "bridge_state": "auto",
  "allow_lan": true,
  "block_when_disconnected": false,
  "auto_connect": false,
  "tunnel_options": {
    "openvpn": {
      "mssfix": null
    },
    "wireguard": {
      "mtu": null,
      "rotation_interval": {
          "secs": 86400,
          "nanos": 0
      }
    },
    "generic": {
      "enable_ipv6": false
    }
  },
  "settings_version": 3
}
"#;

    #[test]
    fn test_v2_migration() {
        let mut old_settings = serde_json::from_str(V2_SETTINGS).unwrap();

        assert!(version_matches(&mut old_settings));

        migrate(&mut old_settings).unwrap();
        let new_settings: serde_json::Value = serde_json::from_str(V3_SETTINGS).unwrap();

        assert_eq!(&old_settings, &new_settings);
    }
}
