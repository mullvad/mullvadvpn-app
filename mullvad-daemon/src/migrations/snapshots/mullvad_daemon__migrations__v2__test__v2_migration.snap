---
source: mullvad-daemon/src/migrations/v2.rs
expression: "serde_json::to_string_pretty(&settings).unwrap()"
---
{
  "account_token": "1234",
  "allow_lan": true,
  "auto_connect": false,
  "block_when_disconnected": false,
  "bridge_settings": {
    "normal": {
      "location": "any"
    }
  },
  "bridge_state": "auto",
  "relay_settings": {
    "normal": {
      "location": {
        "only": {
          "country": "se"
        }
      },
      "openvpn_constraints": {
        "port": {
          "only": 53
        },
        "protocol": {
          "only": "udp"
        }
      },
      "tunnel_protocol": "any",
      "wireguard_constraints": {
        "port": "any"
      }
    }
  },
  "settings_version": 3,
  "tunnel_options": {
    "generic": {
      "enable_ipv6": false
    },
    "openvpn": {
      "mssfix": null
    },
    "wireguard": {
      "mtu": null,
      "rotation_interval": {
        "nanos": 0,
        "secs": 86400
      }
    }
  }
}
